### PR TITLE
Revert fwnode changes and add description of fwnode cases

### PR DESCRIPTION
**Pull request by bardliao**
There are only four valid fwnode cases which are
- primary->secondary->-ENODEV
- primary->NULL
- secondary->-ENODEV
- NULL

dev->fwnode should be converted between the 4 cases above no matter
how/when set_primary_fwnode() and set_secondary_fwnode() are called.
Describe it in the code so people will keep it in mind.

Fixes: #2599

### CODE REVIEW

**Comment by bardliao** [comment]
@andy-shev Could you review this PR?

**Review by andy-shev — CHANGES_REQUESTED** [review]
Commit message is unclear. Can you rephrase it, like "While the commit SHA ("DESCRIPTION") describes everything correct in its commit message the change it made does the opposite and original commit SHA ("DESCRIPTION") was fully correct. Thus, revert the former one here and improve documentation in the next patch". Also it misses Fixes tag.

**Review by andy-shev — CHANGES_REQUESTED** [review]
add spaces and one dash in the list, so it will be ' --> ' to avoid potential confusion with C pointers.

**Comment by bardliao** [comment]
Thanks @andy-shev PR updated.

File: drivers/base/core.c [hunk 1]
```diff
@@ -4278,7 +4278,7 @@ void set_primary_fwnode(struct device *dev, struct fwnode_handle *fwnode)
 		if (fwnode_is_primary(fn)) {
 			dev->fwnode = fn->secondary;
 			if (!(parent && fn == parent->fwnode))
-				fn->secondary = ERR_PTR(-ENODEV);
+				fn->secondary = NULL;
```
Comment (lyakh) [inline drivers/base/core.c:4288]: I think by "original commit SHA ("DESCRIPTION") was fully correct" something different was meant, I think @andy-shev was asking to mention the original commit that introduced the code before https://github.com/thesofproject/linux/commit/d5dcce0c414fcbfe4c2037b66ac69ea5f9b3f75c

Comment (bardliao) [reply drivers/base/core.c:4288]: > I think by "original commit SHA ("DESCRIPTION") was fully correct" something different was meant, I think @andy-shev was asking to mention the original commit that introduced the code before [d5dcce0](https://github.com/thesofproject/linux/commit/d5dcce0c414fcbfe4c2037b66ac69ea5f9b3f75c)

Thanks @lyakh for pointing it out.

**Comment by andy-shev** [comment]
Last version looks good enough to me to be sent upstream.
Consult with Pierre and whoever else who approves this and proceed.

**Comment by bardliao** [comment]
SOFCI TEST

**Comment by bardliao** [comment]
@lyakh Does it look good to you?

**Comment by bardliao** [comment]
SOFCI TEST

**Review by lyakh — APPROVED** [review]

**Comment by andy-shev** [comment]
Can you send these upstream, please?

**Comment by bardliao** [comment]
> Can you send these upstream, please?

Yes, done. And I saw your `Reviewed-by` tag. Thanks @andy-shev
